### PR TITLE
refactor fix of github issue 11295

### PR DESCRIPTION
**Pull request by twhy**
re #11295

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -353,7 +354,10 @@ Model.prototype.$__handleSave = function(options, callback) {
       }
       this.constructor.exists(where, optionsWithCustomValues)
         .then(documentExists => {
-          const matchedCount = !documentExists ? 0 : 1;
+          const matchedCount = documentExists ? 1 : 0;
```
Comment (Uzlopak) [inline lib/model.js:357]: I think I could have coded it even smaller, when I originally wrote that line. 

```suggestion
          const matchedCount = !!documentExists;
```

File: lib/model.js [hunk 2]
```diff
@@ -353,7 +354,10 @@ Model.prototype.$__handleSave = function(options, callback) {
...
+          if (matchedCount == 0 && optimisticConcurrency) {
```
Comment (Uzlopak) [inline lib/model.js:358]: matchedCount of 0 is equivalent to 

```suggestion
          if (!matchedCount && optimisticConcurrency) {
```

Comment (Uzlopak) [reply lib/model.js:358]: We could actually rename it from matchedCount to something like "hasMatchedDocument" or so.

**Comment by Uzlopak** [comment]
Just added some nitpicks, which is more my fault not having them applied in the first PR.

On the other hand: I like that you were not demotivated after the initial discussion :)

**Comment by twhy** [comment]
Sorry, I can't agree with your changes, and I have written down the reason. Please review.

The discussion is great, and it helps everyone learns more about the codebase, much better than just fix the issue.

**Comment by Uzlopak** [comment]
Then atleast use triple equals.

**Comment by twhy** [comment]
@Uzlopak Agree.

**Comment by vkarpov15** [comment]
I put together an alternative fix for this in https://github.com/Automattic/mongoose/commit/fdc88a3bf3f622a26fb45800995e7e87c9ae98eb . You're right that the original fix for #11295 worked mostly by accident rather than by design. But this fix also isn't as correct as I'd like. The fundamental problem came down to that `this.$__.version` wasn't getting set when there were no modified paths, which was throwing off optimistic concurrency. I'll close this one, but thank you for digging into this in such detail :+1:

**Comment by twhy** [comment]
No problem. It's great to see a better fix. Also, thank you for maintaining this library.

**Pull request closed without merging**